pkg/ocibundle/tools/overlay_linux.go

```diff
@@ -8,33 +8,217 @@ package tools
+	// options (like ":ro")
+	BarePath string
+
+	// DirToMount is the path of the directory that will actually be passed to
+	// the mount system-call when mounting this overlay
+	DirToMount string
+}
+
+// Mount performs the necessary steps to mount an individual OverlayItem. It
+// acts as a sort of multiplexer, calling the corresponding private method
+// depending on the type of overlay involved. Note that this method does not
+// mount the overlay itself (which may consist of a series of OverlayItems);
+// that happens in OverlaySet.Mount().
+func (overlay *OverlayItem) Mount() error {
+	switch overlay.Kind {
+	case OLKINDDIR:
+		return overlay.mountDir()
+	case OLKINDSQUASHFS:
+		return overlay.mountSquashfs()
+	default:
+		return fmt.Errorf("internal error: unrecognized image type in prepareWritableOverlay() (kind: %v)", overlay.Kind)
+	}
+}
+
+// mountDir is the private method for mounting directory-based OverlayItems.
+// This involves bind-mounting followed by remounting of the directory onto
+// itself. This pattern of bind-mount followed by remount allows application of
+// more restrictive mount flags than are in force on the underlying filesystem.
+func (overlay *OverlayItem) mountDir() error {
+	var err error
+	if len(overlay.DirToMount) < 1 {
+		overlay.DirToMount = overlay.BarePath
+	}
+
+	if err = ensureOverlayDir(overlay.DirToMount, false, 0); err != nil {
+		return fmt.Errorf("error accessing directory %s: %s", overlay.DirToMount, err)
+	}
+
+	sylog.Debugf("Performing identity bind-mount of %q", overlay.DirToMount)
+	if err = syscall.Mount(overlay.DirToMount, overlay.DirToMount, "", syscall.MS_BIND, ""); err != nil {
+		return fmt.Errorf("failed to bind %s: %s", overlay.DirToMount, err)
+	}
+
+	// Best effort to cleanup mount
+	defer func() {
+		if err != nil {
+			sylog.Debugf("Encountered error with current OverlaySet; attempting to unmount %q", overlay.DirToMount)
+			syscall.Unmount(overlay.DirToMount, syscall.MNT_DETACH)
+		}
+	}()
+
+	// Try to perform remount
+	sylog.Debugf("Performing remount of %q", overlay.DirToMount)
+	if err = syscall.Mount("", overlay.DirToMount, "", syscall.MS_REMOUNT|syscall.MS_BIND, ""); err != nil {
+		return fmt.Errorf("failed to remount %s: %s", overlay.DirToMount, err)
+	}
+
+	return nil
+}
+
+// mountSquashfs is the private method for mounting squashfs-based OverlayItems.
+func (overlay *OverlayItem) mountSquashfs() error {
+	var err error
+	squashfuseCmd, err := bin.FindBin("squashfuse")
+	if err != nil {
+		return fmt.Errorf("use of squashfs overlay requires squashfuse to be installed: %s", err)
+	}
+
+	// Even though fusermount is not needed for this step, we shouldn't
+	// do the squashfuse mount unless we have the necessary tools to
+	// eventually unmount it
+	_, err = bin.FindBin("fusermount")
+	if err != nil {
+		return fmt.Errorf("use of squashfs overlay requires fusermount to be installed: %s", err)
+	}
+	sqshfsDir, err := os.MkdirTemp("", "squashfuse-for-oci-overlay-")
```

Comment:
This has the potential to cause a security issue in the future, that we should guard against.

It's similar to, but less serious / guaranteed, than prior CVEs - https://github.com/apptainer/singularity/commit/98ad49a3de9043ea287d65f4d544632ff074e506

`MkdirTemp` is going to create a secure `700` permissions temporary directory. However, when you mount a squashfs directly onto it, the permissions will then be those of the root directory of the squashfs. If these are more open than `700`, other users could access the overlay (if the allow_other FUSE option is employed).

```
10:04 am $ ls -la /tmp | grep mysecuretempdir
drwx------. 1 dtrudg-sylabs dtrudg-sylabs      0 May 24 10:03 mysecuretempdir/

10:04 am $ squashfuse image.sqfs /tmp/mysecuretempdir/

10:05 am $ ls -la /tmp | grep mysecuretempdir
drwxr-xr-x. 20 root          root               0 May 18 09:57 mysecuretempdir/
```

We can either...

* Setup our mount target dirs so that they are all inside a provided directory which is assumed/required to be secure (e.g. the OCI bundle directory). This would require passing more information into the overlay code.
* Create a nested directory in the temporary directory, and perform the mount on it, so the permissions of the tempdir do not change form 700.

At this point the second option is easier. We should move to move toward the first at a later stage, but it can be deferred here.